internal/security/proxy/serviceplugin.go

```diff
@@ -0,0 +1,233 @@
+				s.loggingClient.Infof("successful to get service pluginID of plugin %s for service %s", pluginName, serviceName)
+
+				return data.PluginID, nil
+			}
+		}
+
+		// no plugin matching pluginName found
+		return "", nil
+	default:
+		return "", fmt.Errorf("unable to get service plugins from Kong's API %s with status code %d: %s", servicePluginURL,
+			resp.StatusCode, string(responseBody))
+	}
+}
+
+func (s *Service) deleteServicePlugin(serviceName, pluginID string) error {
+	deletePluginPath := path.Join(ServicesPath, serviceName, PluginsPath, pluginID)
+	deleteServicePluginURL := strings.Join([]string{s.configuration.KongURL.GetProxyBaseURL(), deletePluginPath}, "/")
+
+	req, err := http.NewRequest(http.MethodDelete, deleteServicePluginURL, http.NoBody)
+	if err != nil {
+		return fmt.Errorf("failed to create delete request for %s: %w", deleteServicePluginURL, err)
+	}
+
+	req.Header.Add(Authorization, Bearer+s.bearerToken)
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to delete service plugin for service %s: %w", serviceName, err)
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+
+	switch resp.StatusCode {
+	case http.StatusNoContent:
+		s.loggingClient.Infof("successfully deleted service plugin for service %s", serviceName)
+
+		return nil
+	default:
+		responseBody, err := ioutil.ReadAll(resp.Body)
+		var errMsg string
+		if err != nil {
+			errMsg = fmt.Sprintf("failed to read response body for delete service plugin: %v", err)
+		} else {
+			errMsg = string(responseBody)
+		}
+
+		return fmt.Errorf("unable to delete service plugin from Kong's API %s with status code %d: %s", deleteServicePluginURL,
+			resp.StatusCode, errMsg)
+	}
+}
+
+func (s *Service) retrieveConsulAccessToken() (string, error) {
+	if len(s.configuration.AccessTokenFile) == 0 {
+		return "", errors.New("required AccessTokenFile from configuration is empty")
+	}
+
+	tokenFileAbsPath, err := filepath.Abs(s.configuration.AccessTokenFile)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert tokenFile to absolute path %s: %v", s.configuration.AccessTokenFile, err)
+	}
+
+	if exists := helper.CheckIfFileExists(tokenFileAbsPath); !exists {
+		return "", fmt.Errorf("access token file %s not found", tokenFileAbsPath)
+	}
+
+	fileOpener := fileioperformer.NewDefaultFileIoPerformer()
+	tokenReader, err := fileOpener.OpenFileReader(tokenFileAbsPath, os.O_RDONLY, 0400)
+	if err != nil {
+		return "", fmt.Errorf("failed to open file reader: %v", err)
+	}
+
+	var accessToken map[string]interface{}
+	if err := json.NewDecoder(tokenReader).Decode(&accessToken); err != nil {
+		return "", fmt.Errorf("failed to parse access token JSON data: %v", err)
+	}
+
+	s.loggingClient.Infof("successfully retrieved access token from %s", tokenFileAbsPath)
+
+	return fmt.Sprintf("%s", accessToken["SecretID"]), nil
+}
+
+func (s *Service) createConsulTokenHeaderForServicePlugin(serviceName, pluginName, accessToken string) error {
```

Comment:
Is there any kind of check to make sure we don't for some reason add the same header twice?

Reply:
there is a deleting before this call, so it won't have the addition of the same header twice